crates/re_sdk/src/recording_stream.rs

```diff
@@ -0,0 +1,1014 @@
+    ///
+    /// Internally, incoming [`DataRow`]s are automatically coalesced into larger [`DataTable`]s to
+    /// optimize for transport.
+    #[inline]
+    pub fn record_row(&self, row: DataRow) {
+        let Some(this) = &*self.inner else {
+            re_log::debug!("Recording disabled - call to record_row() ignored");
+            return;
+        };
+
+        this.batcher.push_row(row);
+    }
+
+    /// Swaps the underlying sink for a new one.
+    ///
+    /// This guarantees that:
+    /// 1. all pending rows and tables are batched, collected and sent down the current sink,
+    /// 2. the current sink is flushed if it has pending data in its buffers,
+    /// 3. the current sink's backlog, if there's any, is forwarded to the new sink.
+    ///
+    /// When this function returns, the calling thread is guaranteed that all future record calls
+    /// will end up in the new sink.
+    ///
+    /// ## Data loss
+    ///
+    /// If the current sink is in a broken state (e.g. a TCP sink with a broken connection that
+    /// cannot be repaired), all pending data in its buffers will be dropped.
+    pub fn set_sink(&self, sink: Box<dyn LogSink>) {
+        let Some(this) = &*self.inner else {
+            re_log::debug!("Recording disabled - call to set_sink() ignored");
+            return;
+        };
+
+        // NOTE: Internal channels can never be closed outside of the `Drop` impl, all these sends
+        // are safe.
+
+        // 1. Flush the batcher down the table channel
+        this.batcher.flush_blocking();
+
+        // 2. Receive pending tables from the batcher's channel
+        this.cmds_tx.send(Command::PopPendingTables).ok();
+
+        // 3. Swap the sink, which will internally make sure to re-ingest the backlog if needed
+        this.cmds_tx.send(Command::SwapSink(sink)).ok();
+
+        // 4. Before we give control back to the caller, we need to make sure that the swap has
+        //    taken place: we don't want the user to send data to the old sink!
+        let (cmd, oneshot) = Command::flush();
+        this.cmds_tx.send(cmd).ok();
+        oneshot.recv().ok();
+    }
+
+    /// Initiates a flush of the pipeline and returns immediately.
+    ///
+    /// This does **not** wait for the flush to propagate (see [`Self::flush_blocking`]).
+    /// See [`RecordingStream`] docs for ordering semantics and multithreading guarantees.
+    pub fn flush_async(&self) {
+        let Some(this) = &*self.inner else {
+            re_log::debug!("Recording disabled - call to flush_async() ignored");
+            return;
+        };
+
+        // NOTE: Internal channels can never be closed outside of the `Drop` impl, all these sends
+        // are safe.
+
+        // 1. Synchronously flush the batcher down the table channel
+        //
+        // NOTE: This _hash_ to be done synchronously as we need to be guaranteed that all tables
+        // are ready to be trained by the time this call returns.
```

Comment:
```suggestion
        // NOTE: This _has_ to be done synchronously as we need to be guaranteed that all tables
        // are ready to be drained by the time this call returns.
```

Reply:
Oh man... I really needed some sleep by that point :joy: